server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java

```diff
@@ -580,6 +610,65 @@ public void syncDeleteDataFiles() {
     }
   }
 
+  /**
+   * Iterate each TsFile and try to lock and remove those out of TTL.
+   */
+  public synchronized void checkFilesTTL() {
+    if (dataTTL == Long.MAX_VALUE) {
+      logger.debug("{}: TTL not set, ignore the check", storageGroupName);
+      return;
+    }
+    long timeBound = System.currentTimeMillis() - dataTTL;
+    if (logger.isDebugEnabled()) {
+      logger.debug("{}: TTL removing files before {}", storageGroupName, new Date(timeBound));
+    }
+    // copy to avoid concurrent modification of deletion
+    List<TsFileResource> seqFiles = new ArrayList<>(sequenceFileList);
+    List<TsFileResource> unseqFiles = new ArrayList<>(unSequenceFileList);
+
+    for (TsFileResource tsFileResource : seqFiles) {
+      checkFileTTL(tsFileResource, timeBound, true);
+    }
+    for (TsFileResource tsFileResource : unseqFiles) {
+      checkFileTTL(tsFileResource, timeBound, false);
+    }
```

Comment:
Why not use Iterator.remove() to avoid concurrent modification of deletion? It can also save the extra list copy time and space.

Reply:
![image](https://user-images.githubusercontent.com/23610645/67254066-96523400-f4ad-11e9-9a0f-b032573e1e31.png)
Unfortunately, the Iterator of ArrayList does not provide concurrent safety, either.